app/controllers/admin/applications_controller.rb

```diff
@@ -22,4 +22,21 @@ def approve
     end
     redirect_to admin_user_path(params[:user_id])
   end
+
+  def mark_paid
+    @application = Application.find(params[:id])
+    if !@application.completed? or @application.payment.has_paid
+      flash[:alert] = @application.completed? == false ? "The application must be completed by the user before it can be marked as paid." : "The user has already paid for this application."
+    else
+      @application.payment.has_paid = true
```

Comment:
I think you should make a separate method to populate the attributes instead of making this method do everything.